events/janus_wsevh.c

```diff
@@ -575,60 +576,55 @@ json_t *janus_wsevh_handle_request(json_t *request) {
 		}
 }
 
-/* Thread to implement the WebSockets loop */
+#if (LWS_LIBRARY_VERSION_MAJOR >= 3 && LWS_LIBRARY_VERSION_MINOR >= 2) || (LWS_LIBRARY_VERSION_MAJOR >= 4)
+/*
+* Websocket thread loop for websocket library newer than 3.2
+* The reconnect is handled in a dedicated lws scheduler janus_wsevh_schedule_connect_attempt
+*/
 static void *janus_wsevh_thread(void *data) {
-	JANUS_LOG(LOG_VERB, "Joining WebSocketsEventHandler client thread\n");
+	JANUS_LOG(LOG_VERB, "Joining WebSocketsEventHandler (lws>=3.2) client thread\n");
+	int n = 0;
+	while(n >= 0 && g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping))
+		n = lws_service(context, 0);
```

Comment:
What can cause `lws_service` to return a negative integer? I'm thinking of intermittent errors that could cause this loop to break before Janus is shutdown.

Reply:
It´s copy pasted from the sample actually.

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/ws-client/minimal-ws-client/minimal-ws-client.c#n210

I will check inside libwebsocket if it returns a different value and what might be conditions

Reply:
If you follow the functions through the lib there are some places where the function returns -1,
I did not validate them all but to me it looks like those are scenarios where the context is not properly settled and therefore further execution does not make sense...

* lws_service() returns 1 or lws_plat_service()
* lws_plat_service() returns _lws_plat_service_tsi()
* _lws_plat_service_tsi() returns on seldom cases -1 but the main negative returns are here lws_service_fd_tsi()

https://libwebsockets.org/git/libwebsockets/tree/lib/core-net/service.c?h=v4.2-stable#n624

The question is whats better in the end.
A 100% thread or a thread termination :)

Reply:
The problem is that this could result in a fatal scenario due to a temporary issue: as a first integration of the return value, I'd just print the negative value in case of errors, but I'd move on with the loop (so no `n>=0` in the `while`). This would give us some more information from people's logs, in case it ever happens.

Reply:
I can add it but i fear i causes more trouble than it solves.
What if the lws_service just returns instantly without sleeping? 
That would just generate endless log entries with one core just writing log entries...